src/ts/lib/safe.ts

```diff
@@ -0,0 +1,168 @@
+import {
+  BigNumber,
+  BigNumberish,
+  BytesLike,
+  constants,
+  Contract,
+  utils,
+} from "ethers";
+import { HardhatEthersHelpers } from "@nomiclabs/hardhat-ethers/types";
+
+export enum SafeOperation {
+  Call = 0,
+  DelegateCall = 1,
+}
+
+export interface SafeDeploymentAddresses {
+  singleton: string;
+  factory: string;
+  fallbackHandler: string;
+}
+
+const gnosisSafeIface = new utils.Interface(GnosisSafe.abi);
+const proxyFactoryIface = new utils.Interface(GnosisSafeProxyFactory.abi);
+const createCallIface = new utils.Interface(CreateCall.abi);
+
+export function safeSetupData(
+  owners: string[],
+  threshold: number,
+  fallbackHandler?: string,
+): string {
+  return gnosisSafeIface.encodeFunctionData("setup", [
+    owners,
+    threshold,
+    constants.AddressZero,
+    "0x",
+    fallbackHandler ?? constants.AddressZero,
+    constants.AddressZero,
+    constants.Zero,
+    constants.AddressZero,
+  ]);
+}
+
+export async function prepareDeterministicSafeWithOwners(
+  owners: string[],
+  threshold: number,
+  {
+    singleton,
+    factory,
+    fallbackHandler,
+  }: Pick<SafeDeploymentAddresses, "singleton" | "factory"> &
+    Partial<SafeDeploymentAddresses>,
+  nonce: BigNumberish,
+  ethers: HardhatEthersHelpers,
+): Promise<{ to: string; data: string; address: string }> {
+  const setupOwnersBytecode = safeSetupData(owners, threshold, fallbackHandler);
+  const proxyFactory = new Contract(
+    factory,
+    GnosisSafeProxyFactory.abi,
+  ).connect(ethers.provider);
+  const createProxyInput = [
+    singleton,
+    setupOwnersBytecode,
+    BigNumber.from(nonce),
+  ];
+  const data: string = proxyFactoryIface.encodeFunctionData(
+    "createProxyWithNonce",
+    createProxyInput,
+  );
+  const address = await proxyFactory.callStatic.createProxyWithNonce(
+    ...createProxyInput,
+  );
+  return { to: proxyFactory.address, data, address };
+}
+
+export function prepareSafeWithOwners(
+  owners: string[],
+  threshold: number,
+  {
+    singleton,
+    factory,
+    fallbackHandler,
+  }: Pick<SafeDeploymentAddresses, "singleton" | "factory"> &
+    Partial<SafeDeploymentAddresses>,
+): { to: string; data: string } {
+  const setupOwnersBytecode = safeSetupData(owners, threshold, fallbackHandler);
+  const data: string = proxyFactoryIface.encodeFunctionData("createProxy", [
+    singleton,
+    setupOwnersBytecode,
+  ]);
+  return { to: factory, data };
+}
+
+export async function createdProxies(
+  response: TransactionResponse,
+  proxyFactoryAddress: string,
+): Promise<string[]> {
+  const receipt = await response.wait();
+  const creationEvents = receipt.logs
+    .filter(({ address }) => address === proxyFactoryAddress)
+    .map((log) => proxyFactoryIface.parseLog(log))
+    .filter(({ name }) => name === "ProxyCreation");
+  return creationEvents.map(({ args }) => args.proxy);
+}
+
+export async function getFallbackHandler(
```

Comment:
This is hard to review. Did you copy the code from somewhere? maybe you can provide a source

Reply:
I added a comment with some references and made the byte decoding more readable. Do you think it makes sense now?

Reply:
Yes, much much better! Now it's easy to understand. Looks good

It's a small pity, that `0x6c9a6c4a39284e37ed1cf53d337577d14212a4870fb976a4366c693b939918d5` was not exported somewhere in their project.